File: macspoof.py
#!/usr/bin/env python3
import subprocess
from getopt import getopt, GetoptError
from os import geteuid
from random import randrange
from re import match
from sys import argv, exit
from time import sleep

version = "macspoof v2.3 (beta) by Max Zinkus <maxzinkus@gmail.com>"

def macspoof():
    try:
        opts = getopt(argv[1:], "hVqi:m:", 
                ['help', 'version', 'quiet', 'interface=', 'mac='])[0]
    except GetoptError as err:
        print(str(err).capitalize() + ".")
        print("Do macspoof -h for usage information.")
        return 1
    interface = None
    spoof_addr = None
    quiet = False
    for opt, arg in opts:
        if opt == '-h' or opt == '--help':
            print(helppage)
            return 0
        elif opt == '-V' or opt == '--version':
            print(version)
            return 0
        elif opt == '-q' or opt == '--quiet':
            quiet = True
        elif opt == '-i' or opt == '--interface':
            interface = arg
        elif opt == '-m' or opt == '--mac':
            spoof_addr = arg
    if subprocess.check_output(['uname', '-s']).strip() != b'Darwin' and not quiet:
        print("Warning: Non-Darwin operating system detected, proceed with caution.")
    if not interface:
        if not quiet:
            print("No interface specified.")
            print("Do macspoof -h for usage information.")
        return 1
    if geteuid() != 0:
        if not quiet:
            print("macspoof needs to be run as root.")
        return 1
    if not validate_interface(interface):
        if not quiet:
            print("Invalid interface.")
        return 1
    if spoof_addr is not None:
        if not validate_addr(spoof_addr):
            if not quiet:
                print("Invalid address given.")
                print("Do macspoof -h for usage information.")
            return 1
    else:
        spoof_addr = generate_mac()
    startmac = subprocess.check_output(['ifconfig', interface, 'ether'])[-19:-2]
    sleep(0.5)
    subprocess.call(['ifconfig', interface, 'ether', spoof_addr])
    sleep(0.5)
    subprocess.call(['ifconfig', interface, 'down'])
    sleep(0.5)
    subprocess.call(['ifconfig', interface, 'up'])
    sleep(0.5)
    if startmac == subprocess.check_output(['ifconfig', interface, 'ether'])[-19:-2]:
        if not quiet:
            print("NIC refused update, you may wish to try again.")
        return 1
    return 0

def validate_interface(interface):
    pattern = "^([0-9a-zA-Z]){3}$"
    if not match(pattern, interface):
        return False
    try:
        subprocess.check_call(['ifconfig', '--', interface],
                stdout=subprocess.DEVNULL, stderr=subprocess.STDOUT)
    except subprocess.CalledProcessError:
        return False
    return True

def validate_addr(addr):
    pattern = "([0-9A-Fa-f]{2}[:]){5}([0-9A-Fa-f]{2})"
    if match(pattern, addr):
        return True
    return False

def generate_mac():
    mac = []
    while len(mac) <= 16:
        byte = "%x" % randrange(16)
        mac.append(byte)
        if len(mac) in [2, 5, 8, 11, 14]:
            mac.append(':')
    return ''.join(mac)

helppage = """NAME
    macspoof - Spoof your Media Access Control address

SYNOPSIS
    As root (EUID 0):
    macspoof -i <id> [-m <mac>]

DESCRIPTION
    This tool changes the Media Access Control (MAC) address reported by your 
    Network Interface Controller (NIC) until the NIC resets (i.e. system 
    reboot). If no MAC address is specified with -m or --mac, a random
    address will be generated and used. While the NIC reads this address out
    of ROM at boot, it can be spoofed by reassignment such as that supported
    in some ifconfig implementations. Note that upon success, data link layer
    connections will likely need to be renegotiated. This tool needs to be run
    as root (unless any of -h, --help, -V, or --version are passed) in order
    to configure the NIC via ifconfig. This software is in beta. Further, it is
    designed for and has only be tested on Mac OS X (specifically >=10.8).
    Results may vary by system. Please do not use this in a manner inconsistent
    with any policy or law, including but not limited to circumvention of MAC
    address filtering. No warranty whatsoever is offered for this software.

OPTIONS
    -h, --help                              Print usage info and exit
    -V, --version                           Print version info and exit

    
    -i, --interface <id>    Interface identifier (See below: INTERFACE INFO)
    -m, --mac <mac>         Spoof to given address (See below: MAC ADDRESS INFO)

    -q, --quiet             Suppress error messages

INTERFACE INFO
    On most *nix systems, ifconfig will output information about NIC
    interfaces. Here, interface identifier is the ifconfig name of a NIC
    interface. For example, on most Mac OS X systems, the wifi interface is
    ``en1'', and on many Linux systems the wifi interface is ``wlan0'' or
    ``en0''.

MAC ADDRESS INFO
    A valid MAC address consists of six colon-delimited hexadecimal bytes, for
    example:
        50:ca:18:8a:11:de
        de:ad:be:ef:ca:fe
    Capitalization of non-digit characters in hex digits is system-specific,
    however in testing, ifconfig handles the disparity. If none is specified,
    macspoof will generate a random valid address to spoof to.

AUTHOR
    Max Zinkus <maxzinkus@gmail.com>
"""

if __name__ == '__main__':
    exit(macspoof())

File: macspoof.c
#include <unistd.h> 
#include <stdio.h> 
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <regex.h>
#include <time.h>
#include <stdbool.h>

#define ADDR_BYTES 18
#define BYTE_MAX 256
#define MAX_INPUT 64

const char* version = "macspoof v0.1.7 (alpha) by Max Zinkus <maxzinkus@gmail.com>";

void help_message() {
   printf("-i <id> to specify NIC id of chosen interface, \
         and optionally -m <mac> to specify a colon-delimited \
         (':') mac address to spoof to.\n");
}

char *generate_mac() {
   int i;
   char *new_addr = calloc(1, ADDR_BYTES);
   
   srand(time(NULL)); /* Seed the rng */
   
   for (i = 0; i < ADDR_BYTES-3; i+=3) {
      sprintf(new_addr + i, "%0x", rand() % BYTE_MAX);
      new_addr[i+2] = ':';
   }
   sprintf(new_addr + i, "%0x", rand() % BYTE_MAX);
   return new_addr;
}

bool validate_interface(char *id) {
   regex_t idpat;
   char* idpattern = "^[a-zA-Z0-9]+$";
   char *test_command;

   regcomp(&idpat, idpattern, REG_EXTENDED);

   if (regexec(&idpat, id, 0, NULL, 0) == REG_NOMATCH) {
      return false;
   }
   else {
      test_command = calloc(1, strlen("ifconfig ") + strnlen(id, MAX_INPUT) +
                            strlen(" >/dev/null")+1
                           );
      sprintf(test_command, "%s %s %s", "ifconfig", id, ">/dev/null");
      
      /* checks if it's actually valid
       * returns 0 => returns true; returns 1 => returns false
       */
      return ! (bool) system(test_command); 
   }
}

bool validate_address(char *addr) {
   regex_t addrpat;
   char* addrpattern = "([0-9A-Fa-f]{2}[:]){5}([0-9A-Fa-f]{2})";

   regcomp(&addrpat, addrpattern, REG_EXTENDED);

   if (regexec(&addrpat, addr, 0, NULL, 0) == REG_NOMATCH) {
      return false;
   }
   else {
      return true;
   }
}

void spoof_interface_mac(char* id, char* spoof_addr) {
   char *command;
   command = calloc(1, strlen("ifconfig ") + strnlen(id, MAX_INPUT) + strlen(" ether ") +
                    strnlen(spoof_addr, ADDR_BYTES)+1
                   );
   sprintf(command, "%s %s %s %s", "ifconfig", id, "ether", spoof_addr);
   system(command);
}

void bounce_interface(char *id) {
   char command[13] = "ifconfig ";
   char up_command[16];
   char down_command[18];
   
   strncat(command, id, sizeof(command) - strlen(command)-1);
   strncpy(up_command, command, sizeof(up_command)-1);
   strncpy(down_command, command, sizeof(down_command)-1);
   strncat(up_command, " up", sizeof(up_command) - strlen(up_command)-1);
   strncat(down_command, " down", sizeof(down_command) - strlen(down_command)-1);
   printf("%s\n", down_command);
   system(down_command);
   sleep(1);
   printf("%s\n", up_command);
   system(up_command);
}

int main(int argc, char **argv) {
   char c, *id = NULL, *spoof_addr = NULL;

   opterr = 0;

   while ((c = getopt(argc, argv, "hVi:m:")) != -1) {
      switch(c) {
         case 'h':
            help_message();
            return 0;
         case 'V':
            printf("%s\n", version);
            return 0;
         case 'i':
            if (validate_interface(optarg)) {
               id = calloc(1, strlen(optarg)+1);
               strcpy(id, optarg);
            }
            else {
               fprintf(stderr, "Invalid interface specified after -i.\n");
               return 1;
            }
            break;
         case 'm':
            if (validate_address(optarg)) {
               spoof_addr = calloc(1, strlen(optarg)+1);
               strcpy(spoof_addr, optarg);
            }
            else {
               fprintf(stderr, "Invalid address specified after -m.\n");
               return 1;
            }
            break;
         case '?':
            if (optopt == 'i' || optopt == 'm') {
               fprintf(stderr, "Option -%c requires an argument.\n", optopt);
               return 1;
            }
            else if (isprint(optopt)) {
               fprintf(stderr, "Unknown option -%c.\n", optopt);
               return 1;
            }
            else {
               fprintf(stderr, "Unknown option character `\\x%x'.\n", optopt);
               return 1;
            }
         default:
            abort(); /* if we get here, bad things have happened */
      }
   }
   if (geteuid() != 0) {
      fprintf(stderr, "This tools needs to be run as root.\n");
      return 1;
   }
   if (!id) { 
      fprintf(stderr, "No interface supplied.\n");
      return 1;
   }
   if (!spoof_addr) { 
      spoof_addr = generate_mac();
      spoof_addr[ADDR_BYTES-1] = '\0';
   }
   spoof_interface_mac(id, spoof_addr);
   sleep(2);
   bounce_interface(id);
   sleep(1);
   return 0;
}
